chapters/memory.txt

```diff
@@ -3096,18 +3096,168 @@ ifdef::VK_KHR_device_group[]
-currently reserved for future use.
+tname:VkMemoryMapFlags is a bitmask type for setting a mask of zero or more
+elink:VkMemoryMapFlagBits.
+--
+
+ifdef::VK_EXT_map_memory2[]
+[open,refpage='vkMapMemory2EXT',desc='Map a memory object into application address space',type='protos']
+--
+Alternatively, to retrieve a host virtual address pointer to a region of a
+mappable memory object, call:
+
+include::{generated}/api/protos/vkMapMemory2EXT.txt[]
+
+  * pname:device is the logical device that owns the memory.
+  * pname:pMemoryMapInfo is a pointer to a slink:VkMemoryMapInfoEXT
+    structure describing parameters of the map.
+  * pname:ppData is a pointer to a `void *` variable in which is returned a
+    host-accessible pointer to the beginning of the mapped range.
+    This pointer minus slink:VkMemoryMapInfoEXT::pname:offset must: be
+    aligned to at least
+    slink:VkPhysicalDeviceLimits::pname:minMemoryMapAlignment.
+
+This function behaves identically to flink:vkMapMemory except that it gets
+its parameters via an extensible structure pointer rather than directly as
+function arguments.
+
+include::{generated}/validity/protos/vkMapMemory2EXT.txt[]
+--
+
+[open,refpage='VkMemoryMapInfoEXT',desc='Structure containing parameters of a memory map operation',type='structs']
+--
+The sname:VkMemoryMapInfoEXT structure is defined as:
+
+include::{generated}/api/structs/VkMemoryMapInfoEXT.txt[]
+
+  * pname:sType is the type of this structure.
+  * pname:pNext is `NULL` or a pointer to a structure extending this
+    structure.
+  * pname:memory is the slink:VkDeviceMemory object to be mapped.
+  * pname:offset is a zero-based byte offset from the beginning of the
+    memory object.
+  * pname:size is the size of the memory range to map, or
+    ename:VK_WHOLE_SIZE to map from pname:offset to the end of the
+    allocation.
+  * pname:flags is reserved for future use.
+
+.Valid Usage
+****
+  * pname:memory must: not be currently host mapped
+  * pname:offset must: be less than the size of pname:memory
+  * If pname:size is not equal to ename:VK_WHOLE_SIZE, pname:size must: be
+    greater than `0`
+  * If pname:size is not equal to ename:VK_WHOLE_SIZE, pname:size must: be
+    less than or equal to the size of the pname:memory minus pname:offset
+  * pname:memory must: have been created with a memory type that reports
+    ename:VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT
+ifdef::VK_KHR_device_group[]
+  * pname:memory must: not have been allocated with multiple instances
+endif::VK_KHR_device_group[]
+ifdef::VK_EXT_map_memory_placed[]
+  * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags, the
+    <<features-memoryMapPlaced, pname:memoryMapPlaced>> feature must: be
+    enabled
+  * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags, the
+    pname:pNext chain must: include a slink:VkMemoryMapPlacedInfoEXT
+    structure
+  * If ename:VK_MEMORY_MAP_PLACED_BIT_EXT is set in pname:flags and the
```

Comment:
What's the point of gating this functionality?  What would prevent any implementation style from subtracting the offset from the desired pointer, and mapping the object starting from the beginning?

Reply:
Then it would use more address space than the application expects.

Reply:
Yeah, if you're going to be mapping over an explicit range, the map has to match exactly.  Otherwise you'll stomp things.  And, sadly, most Linux kernel drivers don't support an `mmap()` offset today.  I'd like to see that fixed but, even if I typed the kernel patches today, it would be a year or two before they hit users so we need the feature bit.

Reply:
Okay, and another question, should it be documented how much address space preceding the mapping is used up by vkMapMemory, so we know where and how big to make the reserved mapping?  Would the mapping always just start in the page of the address passed in? If not, there could be a problem w/ MAP_FIXED overwriting preceding pages unexpectedly.

Reply:
The current plan, if we can get away with it, is that it will map exactly the requested range, like `mmap()` does.  The usual mapping system in DRM allows you to map as little or as much as you want at the beginning of the BO, you just can't offset into it.  As long as any extra space the driver needs is placed at the end, it shouldn't be a problem.

If we need to add something like that, we can, but I'd like to avoid it if we can.

Reply:
Ah I see, and I just noticed the offset alignment requirement a few lines down, perfect.